src/workerd/api/tests/pipe-streams-test.js

```diff
@@ -173,6 +173,691 @@ export const pipeThroughJsToJs = {
+        c.enqueue('hello');
+      },
+      cancel() {
+        cancelCalled = true;
+      },
+    });
+    const writable = new WritableStream({
+      write() {
+        throw new Error('boom');
+      },
+    });
+
+    const pipe = readable.pipeTo(writable);
+
+    await rejects(pipe, { message: 'boom' });
+
+    ok(cancelCalled);
+  },
+};
+
+// Test pipeTo error in JS writable does not cancel JS readable when preventCancel = true
+export const pipeToJsToJsErrorWritablePrevent = {
+  async test() {
+    const chunks = [1, 2];
+    let cancelCalled = false;
+    const readable = new ReadableStream({
+      pull(c) {
+        c.enqueue(chunks.shift());
+        if (chunks.length === 0) c.close();
+      },
+      cancel() {
+        cancelCalled = true;
+      },
+    });
+    const writable = new WritableStream({
+      write() {
+        throw new Error('boom');
+      },
+    });
+
+    const pipe = readable.pipeTo(writable, { preventCancel: true });
+
+    await rejects(pipe, { message: 'boom' });
+
+    ok(!cancelCalled);
+
+    const reader = readable.getReader();
+    await reader.read();
+  },
+};
+
+// Test closing JS readable closes JS writable when preventClose = false
+export const pipeToJsToJsCloseReadable = {
+  async test() {
+    let closeCalled = false;
+    const readable = new ReadableStream({
+      start(c) {
+        c.close();
+      },
+    });
+    const writable = new WritableStream({
+      close() {
+        closeCalled = true;
+      },
+    });
+
+    await readable.pipeTo(writable);
+
+    ok(closeCalled);
+  },
+};
+
+// Test closing JS readable does not close JS writable when preventClose = true
+export const pipeToJsToJsCloseReadablePrevent = {
+  async test() {
+    let closeCalled = false;
+    const readable = new ReadableStream({
+      start(c) {
+        c.close();
+      },
+    });
+    const writable = new WritableStream({
+      close() {
+        closeCalled = true;
+      },
+    });
+
+    await readable.pipeTo(writable, { preventClose: true });
+
+    ok(!closeCalled);
+  },
+};
+
+// Test pipeTo from a tee branch
+export const pipeToJsToJsTee = {
+  async test() {
+    const readable = new ReadableStream({
+      start(c) {
+        c.enqueue('hello');
+        c.close();
+      },
+    });
+
+    const output = [];
+    const writable = new WritableStream({
+      write(chunk) {
+        output.push(chunk);
+      },
+    });
+
+    const [branch] = readable.tee();
+
+    await branch.pipeTo(writable);
+
+    strictEqual(output[0], 'hello');
+  },
+};
+
+// Test pipeTo with already-aborted signal (JS to JS)
+export const pipeToJsToJsCancelAlready = {
+  async test() {
+    const signal = AbortSignal.abort(new Error('boom'));
+    let writeCalled = false;
+    const readable = new ReadableStream({
+      start(c) {
+        c.enqueue('hello');
+        c.close();
+      },
+    });
+
+    const writable = new WritableStream({
+      write(chunk) {
+        writeCalled = true;
```

Comment:
Using mocks would make sure we also validate how many times this function is called as well.